### PR TITLE
Add CPU and RAM usage to Metrics

### PR DESCRIPTION
**Pull request by sagar-solana**
#### Problem

No way to track CPU and RAM usage of the validators in a Cluster

#### Summary of Changes

Add CPU and RAM resource usage tracking to metrics. 

- CPU usage is averaged across all validators
- The "max" RAM usage is is reported across all validators.

Here's what it looks like:
image

Can probably add more data(Disk, GPU usage?) to this later.

### CODE REVIEW

File: scripts/cpu-stats.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+cd "$(dirname "$0")"
+
+# shellcheck source=scripts/configure-metrics.sh
+source configure-metrics.sh
+
+cpu_usage=0
+
+update_cpustat() {
+# collect the total cpu usage by subtracting idle usage from 100%
+cpu_usage=$(top -bn1 | grep '%Cpu(s):' | sed "s/.*, *\([0-9.]*\)%* id.*/\1/" | awk '{print 100 - $1}')
```
Comment (mvines) [inline scripts/cpu-stats.sh]: nit: indent by 2

File: scripts/cpu-stats.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+#
+# Reports cpu usage statistics
+#
+set -e
+
+[[ $(uname) == Linux ]] || exit 0
+
+cd "$(dirname "$0")"
+
+# shellcheck source=scripts/configure-metrics.sh
```
Comment (mvines) [inline scripts/cpu-stats.sh]: nit: if you `cd "$(dirname "$0")/.."` instead then this shellcheck goes away

File: scripts/cpu-stats.sh (hunk 1 above)
Comment (mvines) [reply scripts/cpu-stats.sh]: actually `update_cpustat()` can just be inlined in the main while loop it looks like

File: scripts/cpu-stats.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+while true; do
+  update_cpustat
+  report="cpu_usage=$cpu_usage"
+
+  echo "$report"
+  ./metrics-write-datapoint.sh "cpu-stats,hostname=$HOSTNAME $report"
+  sleep 1
+done
+
+exit 1
```
Comment (mvines) [inline scripts/cpu-stats.sh]: unreachable!

File: scripts/cpu-stats.sh (hunk 1 above)
Comment (sagar-solana) [reply scripts/cpu-stats.sh]: I left it like this so I can follow up with RAM and possible GPU stats

Comment (mvines) [reply scripts/cpu-stats.sh]: ah, ok.  Then remove the extra `update_cpustat` call on line 21?

File: scripts/cpu-stats.sh (hunk 3 above)
Comment (sagar-solana) [reply scripts/cpu-stats.sh]: heh that's what I get for copying boilerplate from other scripts.

File: scripts/cpu-stats.sh (hunk 1 above)
Comment (sagar-solana) [reply scripts/cpu-stats.sh]: That initializes it so it doesn't end up sending 0 the first time?

File: scripts/cpu-stats.sh (hunk 2 above)
Comment (sagar-solana) [reply scripts/cpu-stats.sh]: This doesn't working. `net-stat.sh` has the same shellcheck silence

Comment (sagar-solana) [reply scripts/cpu-stats.sh]: Fixed it! :)

File: scripts/system-stats.sh [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+# Reports cpu and ram usage statistics
+#
+set -e
+
+[[ $(uname) == Linux ]] || exit 0
+
+# need to cd like this to avoid #SC1091
+cd "$(dirname "$0")/.."
+source scripts/configure-metrics.sh
+cd scripts
```
Comment (sagar-solana) [inline scripts/system-stats.sh]: @mvines I had to choose between relative-path script invocation in line 30 or just cd back into the scripts dir here. Does this look ok?

Comment (mvines) [reply scripts/system-stats.sh]: I'd just remove `cd scripts` entirely and reference `scripts/metrics-write-datapoint.sh`

File: scripts/system-stats.sh [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+cpu_usage=0
+ram_usage=0
+
+update_cpustat() {
+# collect the total cpu usage by subtracting idle usage from 100%
+cpu_usage=$(top -bn1 | grep '%Cpu(s):' | sed "s/.*, *\([0-9.]*\)%* id.*/\1/" | awk '{print 100 - $1}')
+# collect the total ram usage by dividing used memory / total memory
+ram_total_and_usage=$(top -bn1 | grep 'MiB Mem'| sed "s/.*: *\([0-9.]*\)%* total.*, *\([0-9.]*\)%* used.*/\1 \2/")
+read -r total used <<< "$ram_total_and_usage"
+ram_usage=$(awk "BEGIN {print $used / $total * 100}")
```
Comment (mvines) [inline scripts/system-stats.sh]: indent by 2 spaces please

File: scripts/system-stats.sh [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+while true; do
+  update_cpustat
```
Comment (mvines) [inline scripts/system-stats.sh]: I still think you should inline `update_cpustat()` here, especially given it modifies global variables

File: scripts/system-stats.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+  report="cpu_usage=$cpu_usage,ram_usage=$ram_usage"
+
+  echo "$report"
```
Comment (mvines) [inline scripts/system-stats.sh]: `metrics-write-datapoint.sh` already echos the datapoint information so this additional echo is superfluous.

File: scripts/system-stats.sh (hunk 5 above)
Comment (sagar-solana) [reply scripts/system-stats.sh]: oh sorry. You mentioned it before too. On it.

File: scripts/system-stats.sh (hunk 6 above)
Comment (sagar-solana) [reply scripts/system-stats.sh]: Alright got rid of the fn and pulled it inline.

**Pull request merged**